### DISCUSSION TITLE
Mistakes in GreenALS_SLPdeg_PD FIS file

### DISCUSSION THREAD

**Discussion by juliagrabowski** [Riverscapes/gcd#414] [I need help please]
The GreenALS_SLPdeg_PD FIS file in the default library doesn't run. It defines three categories for slope and point density, but then in the rule set refers to a missing fourth category for slope. It also defines an elevation uncertainty range of [0 0.75], but then the high uncertainty category maxes out at 0.6. I am developing a FIS for a green Lidar dataset and would like to use this ruleset as a reference. I am curious what the author originally intended for the ruleset and if the mistakes could be rectified.

**Comment by joewheaton** [comment]
Hi @juliagrabowski. So, that [FIS ](https://github.com/Riverscapes/fis-dem-error/blob/bb7cc8884eda42b3db2857e7da4d923840a2d58e/BySurveyType/ALS/GreenALS_ZError_SLPdeg_PD.fis) was made a long, long time ago (back in the Matlab DOD version). Jim McKean and I slapped it together for an AGU poster. It was based on the old EARL platform, which is totally out of date. If you want to build one, you might use one of the [other LiDAR's as a starting point](https://github.com/Riverscapes/fis-dem-error/tree/bb7cc8884eda42b3db2857e7da4d923840a2d58e/BySurveyType/ALS). 

The elevation range should actually probably extend further and yes the high value is generally better to make match the upper end of the range (but it should still work). 
https://github.com/Riverscapes/fis-dem-error/blob/bb7cc8884eda42b3db2857e7da4d923840a2d58e/BySurveyType/ALS/GreenALS_ZError_SLPdeg_PD.fis#L30-L36

As we worked more with LiDAR later, we definitely realized how misleading point density can be as it is often a product of overlapping flight lines and more doesn't mean better (often worse with overlap). Anyhow, you might look at:

Schaffrath KR, Belmont P, Wheaton JM. 2015. Landscape-scale geomorphic change detection: Quantifying spatially variable uncertainty and circumventing legacy data issues. Geomorphology 250 : 334–348. DOI: 10.1016/j.geomorph.2015.09.020

Without testing this one, but knowing I got it to work previously in Matlab, I'm wondering what you mean by "doesn't run"? That can be the case if the input rasters you are providing fall outside the specified range? This is often the case for more recent flights. 

If you are doing one for Green LiDAR, I would go back to the drawing board entirely and likely deal with in-channel and out-of-channel separately (e.g. with seperate FIS).

**Reply by juliagrabowski** [reply]
Hello!

I don't think the issue is the data range. I can create a FIS error surface with parameters smaller than my data range without software errors. I double-checked this by creating a FIS file with all of the slope values divided by 100, and it ran. I am pretty certain the issue is with the missing 4th category of slope. When I delete the rules referring to the 4th category (circled in the image), it works. I am using the GCD ArcGIS plug-in, if that makes a difference.

I have been looking at the other ALS FIS files as well, and have some questions about those. The ALS file names are labeled "VegDensity," "VegHeight," and "Veg," though within the file the vegetation components are all labeled VegRoughness. The VegDensity values make sense to me, and I'm guessing are based on NDVI. I am confused by the VegHeight value range, which is -300  to 600. The Veg value range of 0 to 100 makes more sense to me as a measure of vegetation height. I expected the WetDry category to be binary, but ranges in one from 0 to 2 and in the other from 0 to 5. All of this to say... what are the units of these parameters? I haven't been able to find any documentation.

I am intending to incorporate the in-channel / out-of-channel element into the analysis, either with a Wet/Dry binary, or with the ground-check-based accuracy values for wet and dry areas. I hadn't considered creating FIS separately for wet and dry areas -- would you consider that a better option than accounting for differences within the same FIS?

![image](https://user-images.githubusercontent.com/88350850/128261032-10ef9775-3901-40fe-8e30-f855ca1739a8.png)

While I'm at it! I have another question about the GCD ArcGIS plug-in. I am interested in using the Budget Segregation feature with a Directional Mask.  I have created a polygon shapefile following the directions on the website with a field of upstream--> downstream sequentially increasing integer values. When I upload the shapefile to GCD, it can find character fields in the drop down for the "Label Field," but doesn't display my integer field in the drop-down under "Direction Field." The software then rejects the mask. I'm not sure if this is a bug, or if I am missing something. Any advice?

**Reply by fluviotect** [reply]
re: your last question, sounds like something I recently experienced. Philip provides some insight [here](https://github.com/Riverscapes/gcd/issues/402#issuecomment-886835944). 

I posted two workarounds:

- If you're in Arc, then [this workaround](https://github.com/Riverscapes/gcd/issues/402#issuecomment-883823571) may be preferable.

- If using standalone and you've got some python under your belt, it can be addressed via schema settings during the export from geopandas (or fiona). See [this workaround](https://github.com/Riverscapes/gcd/issues/402#issuecomment-890402931).

**Reply by joewheaton** [reply]
Well, you get what you pay for ;).  The best thing to do is figure out how you want your FIS to work, and make your own with your knowledge of the dataset you're working with. We had conversations at some point about having better documentation and a user contributed repository for these. It is still a good idea but funding for GCD dried up, and I stopped pushing and pursing it in favor of other initiatives. It sounds like you're on the right track. Good luck.

**Reply by joewheaton** [reply]
Thanks @fluviotect for those... spot on suggestions for work-arounds. It is annoying, but I guess that's the price we pay for still allowing/working with shapefiles ;).  It is so annoying how many things we work with on GCD where things have been ever so stable, but then it becomes unstable because of some stupid inconsistency or change by ESRI. All in keeping with their commitment to over-charging and under serving all of us ;).

**Reply by fluviotect** [reply]
Amen brother!

**Reply by fluviotect** [reply]
I finally bit the bullet and committed to QGIS and pure python about 9 months ago.  It slowed me down a bit initially, but even at somewhat reduced productivity, I didn't/don't get angry the way I did of the course of 22 years of suffering with ESRI products. Nearly every single time I've dipped back into Arc since, I got reminded of why I've been so eager to get away. 

The only down-side has been that I use shapefiles much more than I had since about 2003, but I look forward to transitioning to sqlite and geopackages...after thesis submission ;-).

**Reply by joewheaton** [reply]
Me too, but only about 5 months ago. I have been surprised how straight-forward transition has been.  The geopackages are amazing! Really powerful and way simpler then the 5-8 files that make up a shapefile. Plus, if you go to [PostGIS](https://riverscapes.xyz/Tools/Technical_Reference/database.html), its an easy transition.

**Reply by fluviotect** [reply]
I've found the transition pretty smooth as well....certainly much easier with versions since 3.10 than when I made earlier attempts at like v2.x and v3.6. Editing was a tough transition and I have some intermittent frustration with layout bugginess, but I really like how I can implement almost anything I can think of with a reasonable amount of effort.  Makes GIS fun again.

Good to hear about geopackages.  I've only encountered when I inadvertently created one while accidentally mixing geometries in geopandas.  Seemed a bit odd as a datatype at first, but easily addressed once I wrapped my head around jumping through the extra (in pure python) hoop. Definitely seems to have good potential.

Cheers for the tip on PostGIS.  Will look into it for sure.  (also after thesis submission)
